src/include/mpitimpl.h

```diff
@@ -1238,5 +1243,83 @@ extern MPID_Thread_mutex_t mpi_t_mutex;
 /* Init and finalize routines */
 extern int MPIR_T_env_init(void);
 extern void MPIR_T_env_finalize(void);
+extern void MPIR_T_events_finalize(void);
+
+typedef MPI_Count(MPIR_T_timestamp_fn) (void);
+
+typedef struct MPIR_T_source_s {
+#ifdef HAVE_ERROR_CHECKING
+    MPIR_T_object_kind kind;
+#endif
+    int index;
+    char *name;
+    char *desc;
+    MPI_T_source_order ordering;
+    MPIR_T_timestamp_fn *timestamp_fn;
+    MPI_Count ticks_per_second;
+    MPI_Count max_ticks;
+
+    UT_hash_handle hh;          /* Makes this structure hashable */
+} MPIR_T_source_t;
+void MPIR_T_register_source(const char *name, const char *desc, MPI_T_source_order ordering,
+                            MPIR_T_timestamp_fn timestamp_fn, MPI_Count ticks_per_second,
+                            MPI_Count max_ticks, int *index);
+
+typedef struct MPIR_T_event_s {
+#ifdef HAVE_ERROR_CHECKING
+    MPIR_T_object_kind kind;
+#endif
+    int index;
+    int source_index;
+    char *name;
+    MPIR_T_verbosity_t verbosity;
+    MPI_Datatype *array_of_datatypes;
+    MPI_Aint *array_of_displacements;
+    int num_elements;
+    MPIR_T_enum_t *enumtype;
+    char *desc;
+    MPIR_T_bind_t bind;
```

Comment:
I may have missed it, where is `MPIR_T_bind_t` defined?

Reply:
It's in `mpi.h` since it contains the definitions of standard user-visible enum values:
```
typedef enum MPIR_T_bind_t {
    /* don't name-shift this if/when MPI_T_ is accepted, this is an MPICH-only
     * extension */
    MPIX_T_BIND_INVALID = 0,

    /* arbitrarily shift values to aid debugging and reduce accidental errors */
    MPI_T_BIND_NO_OBJECT = 9700,
    MPI_T_BIND_MPI_COMM,
    MPI_T_BIND_MPI_DATATYPE,
    MPI_T_BIND_MPI_ERRHANDLER,
    MPI_T_BIND_MPI_FILE,
    MPI_T_BIND_MPI_GROUP,
    MPI_T_BIND_MPI_OP,
    MPI_T_BIND_MPI_REQUEST,
    MPI_T_BIND_MPI_WIN,
    MPI_T_BIND_MPI_MESSAGE,
    MPI_T_BIND_MPI_INFO
} MPIR_T_bind_t;
```

Reply:
Just to be clear, this was a preexisting MPI_T definition.